fetcher/fetcher.go

```diff
@@ -0,0 +1,151 @@
+package fetcher
+
+import (
+	"archive/zip"
+	"context"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"path/filepath"
+
+	"github.com/dcos/dcos-diagnostics/dcos"
+	"github.com/sirupsen/logrus"
+)
+
+// EndpointFetchRequest is a struct passed to Fetcher with information about URL to be fetched
+type EndpointFetchRequest struct {
+	URL      string
+	Node     dcos.Node
+	FileName string
+}
+
+// FetchStatusUpdate is an update message published by Fetcher when EndpointFetchRequest is done. If error occurred	 during
+// fetch then Error field is not nil.
+type FetchStatusUpdate struct {
+	URL   string
+	Error error
+}
+
+// FetchBulkResponse is a message published when Fetcher finish it job due to cancelled context or closed input chanel
+type FetchBulkResponse struct {
+	ZipFilePath string
+}
+
+// Fetcher is a struct responsible for fetching nodes endpoints
+type Fetcher struct {
+	file         *os.File
+	client       *http.Client
+	input        <-chan EndpointFetchRequest
+	statusUpdate chan<- FetchStatusUpdate
+	output       chan<- FetchBulkResponse
+}
+
+// New creates new Fetcher. Fetcher needs to be started with Run()
+func New(
```

Comment:
If I understand correctly, it looks like the fetcher object is created just once whenever we need to generate a diagnostics bundle. If that's the case, couldn't this `New()` function just create the channels and http client itself? Ideally objects like these should be able to be created with just `New()`, instead of having a long list of parameters to pass in.

Reply:
Yeah, currently it's only created once, but the idea is to create ~5 of them to download the endpoints concurrently. And then you need to pass the channels in so they can all put their results in the same channel, etc.